File: LogisimPlus/LogisimPlus/logisim/gui/hex/Clip.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Text;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.gui.hex
{

	using Caret = global::hex.Caret;
	using HexEditor = global::hex.HexEditor;
	using HexModel = global::hex.HexModel;

	internal class Clip : ClipboardOwner
	{
		private static readonly DataFlavor binaryFlavor = new DataFlavor(typeof(int[]), "Binary data");

		private class Data : Transferable
		{
			internal int[] data;

			internal Data(int[] data)
			{
				this.data = data;
			}

			public virtual DataFlavor[] TransferDataFlavors
			{
				get
				{
					return new DataFlavor[] {binaryFlavor, DataFlavor.stringFlavor};
				}
			}

			public virtual bool isDataFlavorSupported(DataFlavor flavor)
			{
				return flavor == binaryFlavor || flavor == DataFlavor.stringFlavor;
			}

// JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
// ORIGINAL LINE: public Object getTransferData(java.awt.datatransfer.DataFlavor flavor) throws UnsupportedFlavorException, java.io.IOException
			public virtual object getTransferData(DataFlavor flavor)
			{
				if (flavor == binaryFlavor)
				{
					return data;
				}
				else if (flavor == DataFlavor.stringFlavor)
				{
					int bits = 1;
					for (int i = 0; i < data.Length; i++)
					{
						int k = data[i] >> bits;
						while (k != 0 && bits < 32)
						{
							bits++;
							k >>= 1;
						}
					}

					int chars = (bits + 3) / 4;
					StringBuilder buf = new StringBuilder();
					for (int i = 0; i < data.Length; i++)
					{
						if (i > 0)
						{
							buf.Append(i % 8 == 0 ? '\n' : ' ');
						}
						string s = Convert.ToString(data[i], 16);
						while (s.Length < chars)
						{
							s = "0" + s;
						}
						buf.Append(s);
					}
					return buf.ToString();
				}
				else
				{
					throw new UnsupportedFlavorException(flavor);
				}
			}
		}

		private HexEditor editor;

		internal Clip(HexEditor editor)
		{
			this.editor = editor;
		}

		public virtual void copy()
		{
			Caret caret = editor.Caret;
			long p0 = caret.Mark;
			long p1 = caret.Dot;
			if (p0 < 0 || p1 < 0)
			{
				return;
			}
			if (p0 > p1)
			{
				long t = p0;
				p0 = p1;
				p1 = t;
			}
			p1++;

			int[] data = new int[(int)(p1 - p0)];
			HexModel model = editor.Model;
			for (long i = p0; i < p1; i++)
			{
				data[(int)(i - p0)] = model.get(i);
			}

			Clipboard clip = editor.getToolkit().getSystemClipboard();
			clip.setContents(new Data(data), this);
		}

		public virtual bool canPaste()
		{
			Clipboard clip = editor.getToolkit().getSystemClipboard();
			Transferable xfer = clip.getContents(this);
			return xfer.isDataFlavorSupported(binaryFlavor);
		}

		public virtual void paste()
		{
			Clipboard clip = editor.getToolkit().getSystemClipboard();
			Transferable xfer = clip.getContents(this);
			int[] data;
			if (xfer.isDataFlavorSupported(binaryFlavor))
			{
				try
				{
					data = (int[]) xfer.getTransferData(binaryFlavor);
				}
				catch (UnsupportedFlavorException)
				{
					return;
				}
				catch (IOException)
				{
					return;
				}
			}
			else if (xfer.isDataFlavorSupported(DataFlavor.stringFlavor))
			{
				string buf;
				try
				{
					buf = (string) xfer.getTransferData(DataFlavor.stringFlavor);
				}
				catch (UnsupportedFlavorException)
				{
					return;
				}
				catch (IOException)
				{
					return;
				}

				try
				{
					data = HexFile.parse(new StringReader(buf));
				}
				catch (IOException e)
				{
					JOptionPane.showMessageDialog(editor.getRootPane(), e.Message, Strings.get("hexPasteErrorTitle"), JOptionPane.ERROR_MESSAGE);
					return;
				}
			}
			else
			{
				JOptionPane.showMessageDialog(editor.getRootPane(), Strings.get("hexPasteSupportedError"), Strings.get("hexPasteErrorTitle"), JOptionPane.ERROR_MESSAGE);
				return;
			}

			Caret caret = editor.Caret;
			long p0 = caret.Mark;
			long p1 = caret.Dot;
			if (p0 == p1)
			{
				HexModel model = editor.Model;
				if (p0 + data.Length - 1 <= model.LastOffset)
				{
					model.set(p0, data);
				}
				else
				{
					JOptionPane.showMessageDialog(editor.getRootPane(), Strings.get("hexPasteEndError"), Strings.get("hexPasteErrorTitle"), JOptionPane.ERROR_MESSAGE);
				}
			}
			else
			{
				if (p0 < 0 || p1 < 0)
				{
					return;
				}
				if (p0 > p1)
				{
					long t = p0;
					p0 = p1;
					p1 = t;
				}
				p1++;

				HexModel model = editor.Model;
				if (p1 - p0 == data.Length)
				{
					model.set(p0, data);
				}
				else
				{
					JOptionPane.showMessageDialog(editor.getRootPane(), Strings.get("hexPasteSizeError"), Strings.get("hexPasteErrorTitle"), JOptionPane.ERROR_MESSAGE);
				}
			}
		}

		public virtual void lostOwnership(Clipboard clip, Transferable transfer)
		{
		}

	}

}

File: logisim/app/src/main/java/logisim/gui/hex/Clip.kt
package logisim.gui.hex

import java.awt.datatransfer.*
import java.io.IOException
import javax.swing.JOptionPane
import hex.Caret
import hex.HexEditor

class Clip(private val editor: HexEditor) : ClipboardOwner {

	companion object {
		private val binaryFlavor = DataFlavor(IntArray::class.java, "Binary data")
	}

	private class Data(private val data: IntArray) : Transferable {
		override fun getTransferDataFlavors(): Array<DataFlavor> =
		arrayOf(binaryFlavor, DataFlavor.stringFlavor)

		override fun isDataFlavorSupported(flavor: DataFlavor): Boolean =
				flavor == binaryFlavor || flavor == DataFlavor.stringFlavor

		@Throws(UnsupportedFlavorException::class, IOException::class)
		override fun getTransferData(flavor: DataFlavor): Any {
			return when (flavor) {
				binaryFlavor -> data
				DataFlavor.stringFlavor -> dataToHexString(data)
                else -> throw UnsupportedFlavorException(flavor)
			}
		}

		private fun dataToHexString(data: IntArray): String {
			val maxBits = data.maxOf { it.toString(2).length }
			val chars = (maxBits + 3) / 4
			return data.joinToString(separator = " ") { it.toString(16).padStart(chars, '0') }
		}
	}

	fun copy() {
		val caret = editor.caret
		val range = getValidRange(caret) ?: return
				val model = editor.model
		val data = IntArray((range.second - range.first + 1).toInt()) { i ->
				model.get(range.first + i)
		}
		editor.toolkit.systemClipboard.setContents(Data(data), this)
	}

	fun canPaste(): Boolean {
		val clipboard = editor.toolkit.systemClipboard
		val content = clipboard.getContents(this)
		return content.isDataFlavorSupported(binaryFlavor)
	}

	fun paste() {
		val clipboard = editor.toolkit.systemClipboard
		val transferable = clipboard.getContents(this) ?: return

				val data: IntArray = try {
			when {
				transferable.isDataFlavorSupported(binaryFlavor) ->
				transferable.getTransferData(binaryFlavor) as IntArray
				transferable.isDataFlavorSupported(DataFlavor.stringFlavor) ->
				parseHexString(transferable.getTransferData(DataFlavor.stringFlavor) as String)
                else -> {
					showError("hexPasteSupportedError", "hexPasteErrorTitle")
					return
				}
			}
		} catch (e: Exception) {
			showError("hexPasteErrorTitle", e.message)
			return
		}

		val caret = editor.caret
		val range = getValidRange(caret)
		if (range == null || !applyPasteData(data, range)) {
			showError("hexPasteErrorTitle", "hexPasteSizeError")
		}
	}

	override fun lostOwnership(clipboard: Clipboard?, contents: Transferable?) {
		// No specific handling required
	}

	private fun getValidRange(caret: Caret): Pair<Long, Long>? {
		var p0 = caret.mark
		var p1 = caret.dot
		if (p0 < 0 || p1 < 0) return null
		if (p0 > p1) p0 = p1.also { p1 = p0 }
		return Pair(p0, p1 + 1)
	}

	private fun applyPasteData(data: IntArray, range: Pair<Long, Long>): Boolean {
		val model = editor.model
		return if (range.first == range.second - 1) {
			if (range.first + data.size - 1 <= model.lastOffset) {
				model.set(range.first, data)
				true
			} else {
				showError("hexPasteErrorTitle", "hexPasteEndError")
				false
			}
		} else {
			if (range.second - range.first == data.size.toLong()) {
				model.set(range.first, data)
				true
			} else {
				false
			}
		}
	}

	private fun parseHexString(hexString: String): IntArray {
		return hexString.lineSequence()
				.flatMap { it.split(Regex("\\s+")) }
            .map { it.toInt(16) }
            .toList()
				.toIntArray()
	}

	private fun showError(titleKey: String, messageKey: String?) {
		JOptionPane.showMessageDialog(editor.rootPane, messageKey, titleKey, JOptionPane.ERROR_MESSAGE)
	}
}
